source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,201 @@
+    private Range r;
+
+    this(Collection c, Range r)
+    {
+        this.c = c;
+        this.r = r;
+    }
+
+    @property auto ref front()
+    {
+        return r.indexed(c.front);
+    }
+
+    void popFront()
+    {
+        c.popFront;
+    }
+
+    @property bool empty()
+    {
+        return c.empty;
+    }
+
+    static if(hasLength!Collection)
+    {
+        @property size_t length()
+        {
+            return c.length;
+        }
+    }
+
+    static if(isForwardRange!Collection && isForwardRange!Range)
+    {
+        @property typeof(this) save()
+        {
+            return IndexedRoR!(Collection, Range)(c.save, r.save);
+        }
+    }
+}
+
+IndexedRoR!(Collection, Range) indexedRoR(Collection, Range)(Collection c, Range r)
+if (isInputRange!Range)
+{
+    return IndexedRoR!(Collection, Range)(c, r);
+}
+
+Permutations permutations(size_t n) @safe pure nothrow
+{
+    assert(n, "must have at least one item");
+    return Permutations(new size_t[n-1], new size_t[n]);
+}
+
+auto permutations(Range)(Range r) @safe pure nothrow
+if (isInputRange!Range && hasLength!Range)
+{
+    auto perms = .permutations(r.length);
+    return IndexedRoR!(Permutations, Range)(perms, r);
+}
+
+///
+pure @safe nothrow unittest
+{
+    import std.algorithm.comparison : equal;
+    import std.range : iota;
+
+    auto expectedRes = [[0, 1, 2],
+         [1, 0, 2],
+         [2, 0, 1],
+         [0, 2, 1],
+         [1, 2, 0],
+         [2, 1, 0]];
+
+    auto r = iota(3);
+    auto rp = permutations(r.length).indexedRoR(r);
+    assert(rp.equal!equal(expectedRes));
+
+    // direct style
+    auto rp2 = iota(3).permutations;
+    assert(rp2.equal!equal(expectedRes));
+}
+
+pure @safe nothrow unittest
+{
+    // is copyable?
+    import std.algorithm: equal;
+    import std.range: iota, dropOne;
+    auto a = iota(2).permutations;
+    assert(a.front.equal([0, 1]));
+    assert(a.save.dropOne.front.equal([1, 0]));
+    assert(a.front.equal([0, 1]));
+}
+
+Permutations makePermutations(Allocator)(auto ref Allocator alloc, size_t n)
+{
+    assert(n, "must have at least one item");
+    import std.experimental.allocator: makeArray;
+    auto state = alloc.makeArray!size_t(n - 1);
+    auto indices = alloc.makeArray!size_t(n);
+    return Permutations(state, indices);
+}
+
+void dispose(Allocator)(auto ref Allocator alloc, auto ref Permutations perm)
+{
+    import std.experimental.allocator: dispose;
+    dispose(alloc, perm.state);
+    dispose(alloc, perm.indices);
+}
+
+static if (__VERSION__ > 2069) @nogc unittest
+{
+    import std.algorithm: equal;
+    import std.range : iota;
+
+    import std.experimental.allocator.mallocator;
+
+    static immutable expected2 = [[0, 1], [1, 0]];
+    auto r = iota(2);
+    auto rp = makePermutations(Mallocator.instance, r.length);
+    assert(rp.indexedRoR(r).equal!equal(expected2));
+    dispose(Mallocator.instance, rp);
+}
+
+struct Permutations
+{
+    private size_t[] indices, state;
```

Comment:
ehm we use a lot of `.length` and sometimes `index i` which are both type `size_t`, thus we need to do casting - should i still use uint?